packages/signers/src/sign-transaction.ts

```diff
@@ -0,0 +1,202 @@
+import { isTransactionPartialSigner, TransactionPartialSigner } from './transaction-partial-signer';
+import { isTransactionSendingSigner, TransactionSendingSigner } from './transaction-sending-signer';
+import { isTransactionSigner, TransactionSigner } from './transaction-signer';
+
+type CompilableTransactionWithSigners = CompilableTransaction<
+    TransactionVersion,
+    IInstructionWithSigners<string, TransactionSigner>
+> &
+    Partial<ITransactionWithSignatures>;
+
+/**
+ * Signs a transaction using any signers that may be stored in IAccountSignerMeta instruction accounts
+ * as well as any signers provided explicitly to this function.
+ * It will ignore TransactionSendingSigners since this function does not send the transaction.
+ */
+export async function partiallySignTransactionWithSigners<
+    TTransaction extends CompilableTransactionWithSigners = CompilableTransactionWithSigners
+>(
+    transaction: TTransaction,
+    signers: readonly TransactionSigner[] = []
+): Promise<TTransaction & ITransactionWithSignatures> {
+    const { partialSigners, modifyingSigners } = categorizeTransactionSigners(
+        deduplicateSigners([...getSignersFromTransaction(transaction).filter(isTransactionSigner), ...signers]),
+        false
+    );
+
+    return signModifyingAndPartialTransactionSigners(transaction, modifyingSigners, partialSigners);
+}
+
+/**
+ * Signs a transaction using any signers that may be stored in IAccountSignerMeta instruction accounts
+ * as well as any signers provided explicitly to this function.
+ * It will assert that the transaction is fully signed before returning.
+ * It will ignore TransactionSendingSigners since this function does not send the transaction.
+ */
+export async function signTransactionWithSigners<
+    TTransaction extends CompilableTransactionWithSigners = CompilableTransactionWithSigners
+>(
+    transaction: TTransaction,
+    signers: readonly TransactionSigner[] = []
+): Promise<TTransaction & IFullySignedTransaction> {
+    const signedTransaction = await partiallySignTransactionWithSigners(transaction, signers);
+    assertTransactionIsFullySigned(signedTransaction);
+    return signedTransaction;
+}
+
+/**
+ * Signs and sends a transaction using any signers that may be stored in IAccountSignerMeta
+ * instruction accounts as well as any signers provided explicitly to this function.
+ * It will identify a single TransactionSendingSigners to use for sending the transaction, if any.
+ * Otherwise, it will send the transaction using the provided fallbackSender.
+ */
+export async function signAndSendTransactionWithSigners<
+    TTransaction extends CompilableTransactionWithSigners = CompilableTransactionWithSigners
+>(
+    transaction: TTransaction,
+    signers: readonly TransactionSigner[] = [],
+    fallbackSender?: (transaction: TTransaction & IFullySignedTransaction) => Promise<SignatureBytes>
```

Comment:
Should we opt for a "fallback" signer or a "default" signer? In other words, do I get to:
- provide the sender I want you to use if none of the signers are senders, or
- provide the sender I want you to use, period?

Reply:
The first statement is true. It will only be used if no `TransactionSendingSigner` was provided.

Reply:
I meant which would be more useful

Reply:
Oh definitely the former. If there is a sending signer and you're not using it, chances are your transaction won't be fully signed.

Additionally, the latter can already be achieved using the `signTransactionWithSigners` function which will ignore any sending signers (that don't implement another interface).

```ts
const signedTransaction = await signTransactionWithSigners(transaction);
await rpc.sendTransaction(signedTransaction).send();
```

See [this related thread](https://github.com/solana-labs/solana-web3.js/pull/1792#discussion_r1394083824).

Reply:
I was thinking someone like a wallet may wish to forcibly use their sender over any others, regardless of whoever else in the transaction is a sender.

However, I think if a wallet wants to ensure they send every transaction, they'd probably use the snippet you provided.

I believe I was also considering a situation where `TransactionSendingSigner` has both `signTransaction` _and_ `signAndSendTransaction`. But that's not the case.

> If there is a sending signer and you're not using it, chances are your transaction won't be fully signed.

This would also happen if you had more than one sender, right? You'd be sending before the next one could sign.

Reply:
> I was thinking someone like a wallet may wish to forcibly use their sender over any others

Interesting, I always see the Signer API being **used** by dApps and **implemented** by wallets but technically wallets could also use the Signer API and create a Signer API inception haha.

But yeah typically wallets would be implementing the `signAndSendTransactions` method themselves and if they use the new web3.js internally, then I think they'd be using lower-level functions from `@solana/keys` to sign and `@solana/rpc-*` to send.

> This would also happen if you had more than one sender, right? You'd be sending before the next one could sign.

That's right. That's the second type of "conflict" documented on Notion:

> ### Handling sender conflicts
> 
> The second type of conflict we’ve identified is when multiple transaction sending signers are provided.
> 
> For instance, say we have two `TransactionSendingSigner` A and B for a given transaction, one using Phantom and another using Solflare. How does one decide which one to use?
> 
> Here as well, **the Signer may allow the app** to decide by implementing the `TransactionSendingSigner` interface in conjunction with the `TransactionPartialSigner` or `TransactionModifyingSigner` interface.
> 
> However, if both signers *must* send the transaction, then there is nothing that can be done (with or without a Signer API) and we should simply choose one which _could_ end up in a transaction failure due to a lack of signer.

The reason it **could** and not **would** end up in a failure is that there can be some valid use-cases where multiple `TransactionSendingSigner` are provided and whichever one is chosen will sign for all other signers (so it doesn't matter which one we choose). An example of that is ephemeral signers.